File: backend/app/services/billing.py
"""
Stripe billing service.

Provides:
  - create_customer()              — create or return existing Stripe customer
  - create_checkout_session()      — create a hosted Stripe Checkout session
  - create_billing_portal()        — create a Stripe Customer Portal session
  - handle_webhook()               — validate sig and dispatch Stripe events
  - get_subscription()             — fetch a Stripe subscription object
  - update_subscription_models()   — store selected model IDs in subscription metadata
  - get_price_id_for_plan()        — map plan name → Stripe price ID
"""

from __future__ import annotations

import uuid
import logging
from datetime import datetime, timezone
from typing import Optional

import stripe
from stripe import StripeError
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import get_settings
from app.models.users import User
from app.models.subscriptions import Subscription
from app.models.billing_logs import BillingLog

logger = logging.getLogger(__name__)

settings = get_settings()
stripe.api_key = settings.STRIPE_SECRET_KEY

PLAN_QUOTAS: dict[str, int] = {
    "free": 100_000,
    "starter": 1_000_000,
    "pro": 5_000_000,
    "team": 20_000_000,
}

PLAN_MAX_MEMBERS: dict[str, int] = {
    "free": 1,
    "starter": 1,
    "pro": 5,
    "team": 25,
}


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


def get_price_id_for_plan(plan: str) -> Optional[str]:
    mapping = {
        "starter": settings.STRIPE_PRICE_ID_STARTER,
        "pro": settings.STRIPE_PRICE_ID_PRO,
        "team": settings.STRIPE_PRICE_ID_TEAM,
    }
    return mapping.get(plan)


async def create_customer(
    db: AsyncSession,
    *,
    user: User,
) -> str:
    """
    Look up an existing Stripe customer ID on the subscription, or create one.
    Returns the Stripe customer ID.
    """
    # Check if subscription already has a customer ID
    result = await db.execute(
        select(Subscription).where(Subscription.user_id == user.id)
    )
    subscription = result.scalar_one_or_none()
    if subscription and subscription.stripe_customer_id:
        return subscription.stripe_customer_id

    if not settings.STRIPE_SECRET_KEY:
        customer_id = "cus_mock_" + user.id.hex[:12]
        if subscription is None:
            subscription = Subscription(
                id=uuid.uuid4(),
                user_id=user.id,
                stripe_customer_id=customer_id,
                plan="free",
                status="active",
                token_quota_monthly=PLAN_QUOTAS["free"],
                token_used_this_month=0,
            )
            db.add(subscription)
        else:
            subscription.stripe_customer_id = customer_id
            subscription.updated_at = _utcnow()
        await db.flush()
        return customer_id

    # Create Stripe customer
    customer = stripe.Customer.create(
        email=user.email,
        name=user.name,
        metadata={"user_id": str(user.id), "clerk_id": user.clerk_id},
    )
    customer_id = customer["id"]

    # Create or update subscription record
    if subscription is None:
        subscription = Subscription(
            id=uuid.uuid4(),
            user_id=user.id,
            stripe_customer_id=customer_id,
            plan="free",
            status="active",
            token_quota_monthly=PLAN_QUOTAS["free"],
            token_used_this_month=0,
        )
        db.add(subscription)
    else:
        subscription.stripe_customer_id = customer_id
        subscription.updated_at = _utcnow()

    await db.flush()
    return customer_id


async def create_checkout_session(
    *,
    customer_id: str,
    plan: str,
    success_url: str,
    cancel_url: str,
) -> dict:
    """Create a Stripe Checkout session for the given plan."""
    if not settings.STRIPE_SECRET_KEY:
        return {
            "checkout_url": success_url,
            "session_id": "sess_mock_" + uuid.uuid4().hex[:12],
        }

    price_id = get_price_id_for_plan(plan)
    if not price_id:
        raise ValueError(f"Unknown plan or price not configured: {plan!r}")

    session = stripe.checkout.Session.create(
        customer=customer_id,
        payment_method_types=["card"],
        line_items=[{"price": price_id, "quantity": 1}],
        mode="subscription",
        success_url=success_url,
        cancel_url=cancel_url,
        metadata={"plan": plan},
        subscription_data={"metadata": {"plan": plan}},
        allow_promotion_codes=True,
        billing_address_collection="auto",
    )
    return {"checkout_url": session["url"], "session_id": session["id"]}


async def create_billing_portal(
    *,
    customer_id: str,
    return_url: str,
) -> str:
    """Create a Stripe Customer Portal session and return the URL."""
    if not settings.STRIPE_SECRET_KEY:
        return return_url

    session = stripe.billing_portal.Session.create(
        customer=customer_id,
        return_url=return_url,
    )
    return session["url"]


async def get_subscription(subscription_id: str) -> dict:
    """Fetch a Stripe subscription object."""
    if not settings.STRIPE_SECRET_KEY or subscription_id.startswith("sub_mock") or subscription_id.startswith("sess_mock"):
        return {
            "id": subscription_id,
            "status": "active",
            "customer": "cus_mock_123",
            "items": {"data": [{"price": {"id": "price_mock"}}]},
            "metadata": {},
        }
    return dict(stripe.Subscription.retrieve(subscription_id))


async def update_subscription_models(
    subscription_id: str,
    models: list[str],
) -> None:
    """Store selected model IDs in Stripe subscription metadata."""
    if not settings.STRIPE_SECRET_KEY or subscription_id.startswith("sub_mock") or subscription_id.startswith("sess_mock"):
        return
    stripe.Subscription.modify(
        subscription_id,
        metadata={"selected_models": ",".join(models)},
    )


async def handle_webhook(
    db: AsyncSession,
    payload: bytes,
    sig: str,
) -> dict:
    """
    Validate the Stripe webhook signature and dispatch to the appropriate handler.
    Returns {"status": "ok", "event_type": ...} or raises on failure.
    """
    try:
        event = stripe.Webhook.construct_event(
            payload, sig, settings.STRIPE_WEBHOOK_SECRET
        )
    except stripe.error.SignatureVerificationError as exc:  # type: ignore[attr-defined]
        raise ValueError(f"Invalid webhook signature: {exc}") from exc
    except Exception as exc:
        raise ValueError(f"Webhook parse error: {exc}") from exc

    event_type: str = event["type"]
    data_obj = event["data"]["object"]
    stripe_event_id: str = event["id"]

    # Idempotency: skip if already processed
    existing = await db.scalar(
        select(BillingLog).where(BillingLog.stripe_event_id == stripe_event_id)
    )
    if existing:
        return {"status": "already_processed", "event_type": event_type}

    user_id: Optional[uuid.UUID] = None

    # Dispatch
    if event_type in (
        "customer.subscription.created",
        "customer.subscription.updated",
        "customer.subscription.deleted",
    ):
        user_id = await _handle_subscription_change(db, data_obj, event_type)

    elif event_type == "checkout.session.completed":
        user_id = await _handle_checkout_completed(db, data_obj)

    elif event_type in (
        "invoice.payment_succeeded",
        "invoice.payment_failed",
    ):
        user_id = await _handle_invoice_event(db, data_obj, event_type)

    # Record the billing log
    if user_id:
        amount = data_obj.get("amount_total") or data_obj.get("amount_due") or 0
        currency = data_obj.get("currency", "usd")
        log = BillingLog(
            id=uuid.uuid4(),
            user_id=user_id,
            stripe_event_id=stripe_event_id,
            event_type=event_type,
            amount_cents=int(amount),
            currency=currency,
            status="processed",
            log_metadata={"object_id": data_obj.get("id")},
            created_at=_utcnow(),
        )
        db.add(log)

    return {"status": "ok", "event_type": event_type}


async def _handle_subscription_change(
    db: AsyncSession,
    data: dict,
    event_type: str,
) -> Optional[uuid.UUID]:
    """Update the local Subscription record when Stripe subscription changes."""
    stripe_sub_id: str = data["id"]
    customer_id: str = data["customer"]
    new_status: str = data["status"]

    # Map Stripe status → plan from metadata or items
    plan = (data.get("metadata") or {}).get("plan", "free")
    if not plan or plan not in PLAN_QUOTAS:
        # Try to infer from price_id
        items = data.get("items", {}).get("data", [])
        if items:
            price_id = items[0].get("price", {}).get("id", "")
            reverse_map = {
                settings.STRIPE_PRICE_ID_STARTER: "starter",
                settings.STRIPE_PRICE_ID_PRO: "pro",
                settings.STRIPE_PRICE_ID_TEAM: "team",
            }
            plan = reverse_map.get(price_id, "free")

    # Find subscription by stripe_customer_id
    result = await db.execute(
        select(Subscription).where(Subscription.stripe_customer_id == customer_id)
    )
    sub = result.scalar_one_or_none()
    if sub is None:
        logger.warning("No subscription found for Stripe customer %s", customer_id)
        return None

    # Update fields
    sub.stripe_subscription_id = stripe_sub_id
    sub.plan = plan if event_type != "customer.subscription.deleted" else "free"
    sub.status = "canceled" if event_type == "customer.subscription.deleted" else new_status
    sub.token_quota_monthly = PLAN_QUOTAS.get(sub.plan, PLAN_QUOTAS["free"])

    # Billing period
    period_start = data.get("current_period_start")
    period_end = data.get("current_period_end")
    if period_start:
        sub.billing_period_start = datetime.fromtimestamp(period_start, tz=timezone.utc)
    if period_end:
        sub.billing_period_end = datetime.fromtimestamp(period_end, tz=timezone.utc)

    sub.updated_at = _utcnow()

    # Sync plan on User record
    user = await db.get(User, sub.user_id)
    if user:
        user.plan = sub.plan
        user.updated_at = _utcnow()

    return sub.user_id


async def _handle_checkout_completed(
    db: AsyncSession,
    data: dict,
) -> Optional[uuid.UUID]:
    """When checkout completes, update the subscription record."""
    subscription_id = data.get("subscription")
    if not subscription_id:
        return None

    stripe_sub = await get_subscription(subscription_id)
    return await _handle_subscription_change(db, stripe_sub, "customer.subscription.created")


async def _handle_invoice_event(
    db: AsyncSession,
    data: dict,
    event_type: str,
) -> Optional[uuid.UUID]:
    """Handle invoice payment success/failure — reset monthly token counter on success."""
    customer_id: str = data.get("customer", "")

    result = await db.execute(
        select(Subscription).where(Subscription.stripe_customer_id == customer_id)
    )
    sub = result.scalar_one_or_none()
    if sub is None:
        return None

    if event_type == "invoice.payment_succeeded":
        # New billing period — reset token usage counter
        sub.token_used_this_month = 0
        sub.status = "active"
        sub.updated_at = _utcnow()

    elif event_type == "invoice.payment_failed":
        sub.status = "past_due"
        sub.updated_at = _utcnow()

    return sub.user_id


class BillingService:
    async def get_subscription(self, db: AsyncSession, user_id: uuid.UUID) -> Optional[Subscription]:
        result = await db.execute(
            select(Subscription).where(Subscription.user_id == user_id)
        )
        return result.scalar_one_or_none()

    async def get_or_create_customer(self, db: AsyncSession, user: User) -> str:
        return await create_customer(db, user=user)

    async def create_checkout_session(
        self,
        customer_id: str,
        price_id: str,
        success_url: str,
        cancel_url: str,
        user_id: Optional[str] = None,
    ) -> str:
        plan = "free"
        if price_id == settings.STRIPE_PRICE_ID_STARTER:
            plan = "starter"
        elif price_id == settings.STRIPE_PRICE_ID_PRO:
            plan = "pro"
        elif price_id == settings.STRIPE_PRICE_ID_TEAM:
            plan = "team"

        res = await create_checkout_session(
            customer_id=customer_id,
            plan=plan,
            success_url=success_url,
            cancel_url=cancel_url,
        )
        return res["checkout_url"]

    async def create_billing_portal(self, customer_id: str, return_url: str) -> str:
        return await create_billing_portal(customer_id=customer_id, return_url=return_url)

    async def update_selected_models(
        self,
        db: AsyncSession,
        user_id: uuid.UUID,
        models: list[str],
    ) -> Subscription:
        result = await db.execute(
            select(Subscription).where(Subscription.user_id == user_id)
        )
        sub = result.scalar_one_or_none()
        if not sub:
            sub = Subscription(
                id=uuid.uuid4(),
                user_id=user_id,
                stripe_customer_id="",
                plan="free",
                status="active",
                selected_models=models,
                token_quota_monthly=PLAN_QUOTAS["free"],
                token_used_this_month=0,
            )
            db.add(sub)
        else:
            sub.selected_models = models
            sub.updated_at = datetime.now(timezone.utc)
            db.add(sub)
        await db.flush()

        if settings.STRIPE_SECRET_KEY and sub.stripe_subscription_id:
            try:
                await update_subscription_models(sub.stripe_subscription_id, models)
            except Exception as exc:
                logger.warning("Failed to sync model selection to Stripe: %s", exc)

        return sub

    async def handle_webhook(self, db: AsyncSession, payload: bytes, sig: str) -> dict:
        return await handle_webhook(db, payload, sig)


billing_service = BillingService()

File: backend/src/services/billing.ts
import Stripe from "stripe";
import { config } from "../config";
import { prisma } from "../prisma";
import crypto from "crypto";

const stripe = new Stripe(config.STRIPE_SECRET_KEY, {
  apiVersion: "2024-04-10" as any, // standard API version
});

export const PLAN_QUOTAS: Record<string, number> = {
  free: 100000,
  starter: 1000000,
  pro: 5000000,
  team: 20000000,
};

export const PLAN_MAX_MEMBERS: Record<string, number> = {
  free: 1,
  starter: 1,
  pro: 5,
  team: 25,
};

export function getPriceIdForPlan(plan: string): string | null {
  const mapping: Record<string, string> = {
    starter: config.STRIPE_PRICE_ID_STARTER,
    pro: config.STRIPE_PRICE_ID_PRO,
    team: config.STRIPE_PRICE_ID_TEAM,
  };
  return mapping[plan] || null;
}

export async function createCustomer(userId: string): Promise<string> {
  const user = await prisma.user.findUnique({
    where: { id: userId },
    include: { subscription: true },
  });

  if (!user) {
    throw new Error("User not found");
  }

  if (user.subscription && user.subscription.stripeCustomerId) {
    return user.subscription.stripeCustomerId;
  }

  if (!config.STRIPE_SECRET_KEY) {
    const customerId = "cus_mock_" + userId.replace(/-/g, "").substring(0, 12);
    if (!user.subscription) {
      await prisma.subscription.create({
        data: {
          userId: user.id,
          stripeCustomerId: customerId,
          plan: "free",
          status: "active",
          tokenQuotaMonthly: PLAN_QUOTAS.free,
          tokenUsedThisMonth: 0,
        },
      });
    } else {
      await prisma.subscription.update({
        where: { id: user.subscription.id },
        data: { stripeCustomerId: customerId },
      });
    }
    return customerId;
  }

  const customer = await stripe.customers.create({
    email: user.email,
    name: user.name,
    metadata: { userId: user.id, clerkId: user.clerkId },
  });

  const customerId = customer.id;

  if (!user.subscription) {
    await prisma.subscription.create({
      data: {
        userId: user.id,
        stripeCustomerId: customerId,
        plan: "free",
        status: "active",
        tokenQuotaMonthly: PLAN_QUOTAS.free,
        tokenUsedThisMonth: 0,
      },
    });
  } else {
    await prisma.subscription.update({
      where: { id: user.subscription.id },
      data: { stripeCustomerId: customerId },
    });
  }

  return customerId;
}

export async function createCheckoutSession(
  customerId: string,
  plan: string,
  successUrl: string,
  cancelUrl: string
): Promise<{ checkout_url: string; session_id: string }> {
  if (!config.STRIPE_SECRET_KEY) {
    return {
      checkout_url: successUrl,
      session_id: "sess_mock_" + crypto.randomBytes(6).toString("hex"),
    };
  }

  const priceId = getPriceIdForPlan(plan);
  if (!priceId) {
    throw new Error(`Unknown plan or price not configured: ${plan}`);
  }

  const session = await stripe.checkout.sessions.create({
    customer: customerId,
    payment_method_types: ["card"],
    line_items: [{ price: priceId, quantity: 1 }],
    mode: "subscription",
    success_url: successUrl,
    cancel_url: cancelUrl,
    metadata: { plan },
    subscription_data: { metadata: { plan } },
    allow_promotion_codes: true,
    billing_address_collection: "auto",
  });

  return {
    checkout_url: session.url || successUrl,
    session_id: session.id,
  };
}

export async function createBillingPortal(customerId: string, returnUrl: string): Promise<string> {
  if (!config.STRIPE_SECRET_KEY) {
    return returnUrl;
  }

  const session = await stripe.billingPortal.sessions.create({
    customer: customerId,
    return_url: returnUrl,
  });

  return session.url;
}

export async function getStripeSubscription(subscriptionId: string): Promise<any> {
  if (!config.STRIPE_SECRET_KEY || subscriptionId.startsWith("sub_mock") || subscriptionId.startsWith("sess_mock")) {
    return {
      id: subscriptionId,
      status: "active",
      customer: "cus_mock_123",
      items: { data: [{ price: { id: "price_mock" } }] },
      metadata: {},
    };
  }
  return await stripe.subscriptions.retrieve(subscriptionId);
}

export async function updateSubscriptionModels(subscriptionId: string, models: string[]): Promise<void> {
  if (!config.STRIPE_SECRET_KEY || subscriptionId.startsWith("sub_mock") || subscriptionId.startsWith("sess_mock")) {
    return;
  }
  await stripe.subscriptions.update(subscriptionId, {
    metadata: { selected_models: models.join(",") },
  });
}

// Sync subscription changes
async function handleSubscriptionChange(data: any, eventType: string): Promise<string | null> {
  const stripeSubId = data.id;
  const customerId = data.customer;
  const newStatus = data.status;

  let plan = data.metadata?.plan || "free";
  if (!plan || !PLAN_QUOTAS[plan]) {
    // Attempt price resolution
    const items = data.items?.data || [];
    if (items.length > 0) {
      const priceId = items[0].price?.id || "";
      const reverseMap: Record<string, string> = {
        [config.STRIPE_PRICE_ID_STARTER]: "starter",
        [config.STRIPE_PRICE_ID_PRO]: "pro",
        [config.STRIPE_PRICE_ID_TEAM]: "team",
      };
      plan = reverseMap[priceId] || "free";
    }
  }

  const sub = await prisma.subscription.findFirst({
    where: { stripeCustomerId: customerId },
  });

  if (!sub) {
    console.warn(`No subscription found for Stripe customer: ${customerId}`);
    return null;
  }

  const periodStart = data.current_period_start ? new Date(data.current_period_start * 1000) : null;
  const periodEnd = data.current_period_end ? new Date(data.current_period_end * 1000) : null;

  const targetPlan = eventType === "customer.subscription.deleted" ? "free" : plan;
  const targetStatus = eventType === "customer.subscription.deleted" ? "canceled" : newStatus;

  await prisma.subscription.update({
    where: { id: sub.id },
    data: {
      stripeSubscriptionId: stripeSubId,
      plan: targetPlan,
      status: targetStatus,
      tokenQuotaMonthly: PLAN_QUOTAS[targetPlan] || PLAN_QUOTAS.free,
      billingPeriodStart: periodStart,
      billingPeriodEnd: periodEnd,
      updatedAt: new Date(),
    },
  });

  await prisma.user.update({
    where: { id: sub.userId },
    data: {
      plan: targetPlan,
      updatedAt: new Date(),
    },
  });

  return sub.userId;
}

async function handleCheckoutCompleted(data: any): Promise<string | null> {
  const subscriptionId = data.subscription;
  if (!subscriptionId) return null;

  const stripeSub = await getStripeSubscription(subscriptionId);
  return await handleSubscriptionChange(stripeSub, "customer.subscription.created");
}

async function handleInvoiceEvent(data: any, eventType: string): Promise<string | null> {
  const customerId = data.customer;

  const sub = await prisma.subscription.findFirst({
    where: { stripeCustomerId: customerId },
  });

  if (!sub) return null;

  if (eventType === "invoice.payment_succeeded") {
    await prisma.subscription.update({
      where: { id: sub.id },
      data: {
        tokenUsedThisMonth: 0,
        status: "active",
        updatedAt: new Date(),
      },
    });
  } else if (eventType === "invoice.payment_failed") {
    await prisma.subscription.update({
      where: { id: sub.id },
      data: {
        status: "past_due",
        updatedAt: new Date(),
      },
    });
  }

  return sub.userId;
}

export async function handleWebhook(payload: any, sig: string): Promise<any> {
  let event: Stripe.Event;

  try {
    event = stripe.webhooks.constructEvent(payload, sig, config.STRIPE_WEBHOOK_SECRET);
  } catch (err: any) {
    throw new Error(`Webhook verification failed: ${err.message}`);
  }

  const eventType = event.type;
  const dataObj: any = event.data.object;
  const stripeEventId = event.id;

  // Idempotency check
  const existingLog = await prisma.billingLog.findUnique({
    where: { stripeEventId },
  });
  if (existingLog) {
    return { status: "already_processed", event_type: eventType };
  }

  let userId: string | null = null;

  if (
    ["customer.subscription.created", "customer.subscription.updated", "customer.subscription.deleted"].includes(
      eventType
    )
  ) {
    userId = await handleSubscriptionChange(dataObj, eventType);
  } else if (eventType === "checkout.session.completed") {
    userId = await handleCheckoutCompleted(dataObj);
  } else if (["invoice.payment_succeeded", "invoice.payment_failed"].includes(eventType)) {
    userId = await handleInvoiceEvent(dataObj, eventType);
  }

  if (userId) {
    const amount = dataObj.amount_total || dataObj.amount_due || 0;
    const currency = dataObj.currency || "usd";

    await prisma.billingLog.create({
      data: {
        userId,
        stripeEventId,
        eventType,
        amountCents: amount,
        currency,
        status: "processed",
        metadata: { object_id: dataObj.id },
      },
    });
  }

  return { status: "ok", event_type: eventType };
}
